### DISCUSSION TITLE
Automatically group peer dependencies

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#21660] [Suggest an Idea]
This feature idea is inspired from suggestions by @ljharb and @Arcanemagus in https://github.com/renovateapp/renovate/issues/323

The idea is to detect when dependencies are related to each other via `peerDependencies` - e.g. `eslint` and `eslint-config-airbnb-base` - and automatically group them together in Renovate PRs without further configuration necessary from the user. 

In the example above it is clearly obvious that a good name for this autogroup is "eslint", however the details are more complex. `eslint-config-airbnb-base` actually has two `peerDependencies`:
```
  "peerDependencies": {
    "eslint": "^4.16.0",
    "eslint-plugin-import": "^2.8.0"
  },
```

How can Renovate "know" what the group name should be? Using this example again with a repo like `renovate`, it should list all three packages as `devDependencies` - `eslint`, `eslint-config-airbnb-base` and `eslint-plugin-import`. It could be detected that `eslint` is listed as a peer by other packages twice, and `eslint-config-import` only once, so we could use that "election" approach. Alternatively, what if two dependencies had "equal" references? Maybe to start with we use a dumb algorithm like whichever's name is shorter, or even which has more downloads?

Complications:

1. What to do if users have defined their own grouping already? I think many have added "^eslint" grouping rules and we don't want to cause *problems*. I suggest we can possibly implement this via an injected `packageRule` and make sure it's first, so that any user-defined packageRules then overwrite it.

2. Our existing "monorepo" groups may conflict with this. e.g. we already have "babel6" and "babel" monorepo groupings which *wouldn't* include third party groupings.

3. What if users don't want this behaviour? e.g. let's say you actually wish for core `angular` packages to be in a monorepo group and third party packages that list `angular` as a peer dependency are upgraded individually?

**Comment by ghost** [comment]
> Maybe to start with we use a dumb algorithm like whichever's name is shorter, or even which has more downloads?

Right now that seems quite reasonable. For ESLint, nyc, Renovate, among a few others, that works well.

However, it won't work out for package peers that don't share a common name.

For example, over in `conventional-changelog`, our convention configuration packages start with `conventional-changelog-`, and they can be a peer package for `conventional-recommended-bump`. Except [we've discussed relaxing the naming requirement](https://github.com/conventional-changelog/conventional-changelog/issues/316) to allow loading any arbitrary named package. Therefore there wouldn't be a common name segment between packages.

**Comment by yangm97** [comment]
What if this was implemented as a "packageRule"? 

```
"packageRules": [
    {
      "matchPeersNames": [
        "react",
        "react-native"
      ],
      "groupName": "react native monorepo",
      "groupSlug": "react-native-monorepo"
    }
  ],
```

This would bundle react, react native and all of their peerDependencies together.

Then I would suggest extending the current presets or even creating a new group of presets for those. Speaking of which, this is most likely the best place to use automation, so we can quickly discover, validate, possibly fix and finally merge peer group suggestions.

I think it would comply with all concerns raised.

**Comment by rarkins** [comment]
I like the idea of letting users control this with `packageRules` as a first step. We'd first need to extend our datasource types to include peer dependencies and add the capability in datasource/npm.  Currently they have dependencies and devDependencies already: https://github.com/renovatebot/renovate/blob/aa4fabe5997e07b2cb4f9f1b5eca8ed9810fc059/lib/datasource/types.ts#L42-L43

Then as you suggest a new option like `matchPeerNames` could be used.

**Comment by RichiCoder1** [comment]
Is this still something people are interested/would take contributions on? I'm interested for the case of `constructs`, `aws-cdk`, and `@serverless-stack` which are all disconnected except A) are all interconnected peer dependencies and B) in the case of serverless-stack is strict about the peer dependency version range.
